### PR TITLE
Add kinematics and odometry docs

### PR DESCRIPTION
**Pull request by prateekma**

### CODE REVIEW

**Comment by prateekma** [comment]
Not sure why linkcheck is failing, I rebased this PR off of master and the only link I added was a local link to another article, which I've verified works.

**Comment by prateekma** [comment]
Now ready for review

File: source/docs/software/kinematics-and-odometry/differential-drive-kinematics.rst [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+=============================
+The ``DifferentialDriveKinematics`` class is a useful tool that converts between a ``ChassisSpeeds`` object and a ``DifferentialDriveWheelSpeeds`` object, which contains velocities for the left and right sides of a differential drive robot.
+
+Constructing the kinematics object
+----------------------------------
+The ``DifferentialDriveKinematics`` object accepts one constructor argument, which is the track width of the robot. This represents the distance between the two sets of wheels on a differential drive.
+
+.. note:: In Java, the track width must be in meters. In C++, the units library can be used to pass in the track width using any length unit.
+
+Converting chassis speeds to wheel speeds
```
Comment (Daltz333) [inline source/docs/software/kinematics-and-odometry/differential-drive-kinematics.rst]:
```suggestion
Converting Chassis Speeds to Wheel Speeds
```

File: source/docs/software/kinematics-and-odometry/differential-drive-kinematics.rst [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      // Example chassis speeds: 2 meters per second linear velocity,
+      // 1 radian per second angular velocity.
+      frc::ChassisSpeeds chassisSpeeds{2_mps, 0_mps, 1_rad_per_s};
+
+      // Convert to wheel speeds. Here, we can use C++17's structured bindings
+      // feature to automatically split the DifferentialDriveWheelSpeeds
+      // struct into left and right velocities.
+      auto [left, right] = kinematics.ToWheelSpeeds(chassisSpeeds);
+
+Converting wheel speeds to chassis speeds
```
Comment (Daltz333) [inline source/docs/software/kinematics-and-odometry/differential-drive-kinematics.rst]:
```suggestion
Converting Wheel Speeds to Chassis Speeds
```

File: source/docs/software/kinematics-and-odometry/differential-drive-odometry.rst [hunk 3]
```diff
@@ -0,0 +1,80 @@
+Differential Drive Odometry
+===========================
+A user can use the differential drive kinematics classes in order to perform odometry. WPILib contains a ``DifferentialDriveOdometry`` class that can be used to track the position of a differential drive robot on the field.
+
+.. note:: Because this method only uses encoders and a gyro, the estimate of the robot's position on the field will drift over time, especially as your robot comes into contact with other robots during gameplay. However, odometry is usually very accurate during the autonomous period.
+
+Creating the odometry object
```
Comment (Daltz333) [inline source/docs/software/kinematics-and-odometry/differential-drive-odometry.rst]:
```suggestion
Creating the Odometry Object
```

File: source/docs/software/kinematics-and-odometry/differential-drive-odometry.rst [hunk 4]
```diff
@@ -0,0 +1,80 @@
+Differential Drive Odometry
+===========================
+A user can use the differential drive kinematics classes in order to perform odometry. WPILib contains a ``DifferentialDriveOdometry`` class that can be used to track the position of a differential drive robot on the field.
```
Comment (Daltz333) [inline source/docs/software/kinematics-and-odometry/differential-drive-odometry.rst]: What is odometry? This needs to be explained. Assume the user knows 0 about autonomous.

File: source/docs/software/kinematics-and-odometry/differential-drive-odometry.rst [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      // Creating my kinematics object: differential drive with track width of 27 inches.
+      frc::DifferentialDriveKinematics m_kinematics{27_in};
+
+      // Creating my odometry object from the kinematics object. Here,
+      // our starting pose is 5 meters along the long end of the field and in the
+      // center of the field along the short end, facing forward.
+      frc::DifferentialDriveOdometry m_odometry{m_kinematics, frc::Pose2d{5_m, 13.5_m, 0_rad}};
+
+
+Updating the robot pose
```
Comment (Daltz333) [inline source/docs/software/kinematics-and-odometry/differential-drive-odometry.rst]:
```suggestion
Updating the Robot Pose
```

File: source/docs/software/kinematics-and-odometry/intro-and-chassis-speeds.rst [hunk 6]
```diff
@@ -0,0 +1,68 @@
+Introduction to Kinematics and The Chassis Speeds Class
+==================================================================
+
+The brand new kinematics suite contains classes for differential drive, swerve drive, and mecanum drive kinematics and odometry. The kinematics classes help convert between a universal ``ChassisSpeeds`` object, containing linear and angular velocities for a robot to usable speeds for each individual type of drivetrain i.e. left and right wheel speeds for a differential drive, four wheel speeds for a mecanum drive, or individual module states (speed and angle) for a swerve drive.
+
+The odometry classes utilize the kinematics classes to track the position of the robot on the field using measurements from encoders and a gyroscope.
+
+The Chassis Speeds class
```
Comment (Daltz333) [inline source/docs/software/kinematics-and-odometry/intro-and-chassis-speeds.rst]:
```suggestion
The Chassis Speeds Class
```

File: source/docs/software/kinematics-and-odometry/intro-and-chassis-speeds.rst [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+   .. code-tab:: c++
+
+      // The robot is moving at 3 meters per second forward, 2 meters
+      // per second to the right, and rotating at half a rotation per
+      // second counterclockwise.
+      frc::ChassisSpeeds speeds{3.0_mps, -2.0_mps,
+        units::radians_per_second_t(wpi::math::pi)};
+
+
+Creating a ChassisSpeeds object from field-relative speeds
```
Comment (Daltz333) [inline source/docs/software/kinematics-and-odometry/intro-and-chassis-speeds.rst]:
```suggestion
Creating a ChassisSpeeds Object from Field-Relative Speeds
```

File: source/docs/software/kinematics-and-odometry/mecanum-drive-kinematics.rst [hunk 8]
```diff
@@ -0,0 +1,139 @@
+Mecanum Drive Kinematics
+========================
+The ``MecanumDriveKinematics`` class is a useful tool that converts between a ``ChassisSpeeds`` object and a ``MecanumDriveWheelSpeeds`` object, which contains velocities for each of the four wheels on a mecanum drive.
+
+Constructing the kinematics object
```
Comment (Daltz333) [inline source/docs/software/kinematics-and-odometry/mecanum-drive-kinematics.rst]:
```suggestion
Constructing the Kinematics Object
```

File: source/docs/software/kinematics-and-odometry/mecanum-drive-kinematics.rst [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+      frc::Translation2d m_frontRightLocation{0.381_m, -0.381_m};
+      frc::Translation2d m_backLeftLocation{-0.381_m, 0.381_m};
+      frc::Translation2d m_backRightLocation{-0.381_m, -0.381_m};
+
+      // Creating my kinematics object using the wheel locations.
+      frc::MecanumDriveKinematics m_kinematics{
+        m_frontLeftLocation, m_frontRightLocation, m_backLeftLocation,
+        m_backRightLocation};
+
+Converting chassis speeds to wheel speeds
```
Comment (Daltz333) [inline source/docs/software/kinematics-and-odometry/mecanum-drive-kinematics.rst]:
```suggestion
Converting Chassis Speeds to Wheel Speeds
```

File: source/docs/software/kinematics-and-odometry/differential-drive-odometry.rst (hunk 4 above)
Comment (prateekma) [reply source/docs/software/kinematics-and-odometry/differential-drive-odometry.rst]: A more detailed explanation of what odometry is was explained in the introductory article. I can add a link to that article here though.

**Pull request merged**